repose-aggregator/components/filters/valkyrie-authorization/src/main/scala/org/openrepose/filters/valkyrieauthorization/ValkyrieAuthorizationFilter.scala

```diff
@@ -337,23 +333,47 @@ class ValkyrieAuthorizationFilter @Inject()(configurationService: ConfigurationS
     }
 
     def cullJsonArray(jsonArray: Seq[JsValue], devicePath: DevicePath, devicePermissions: Vector[DeviceToPermission]): Seq[JsValue] = {
-      jsonArray.filter { value =>
-        val deviceValue: String = Try(JSONPath.query(devicePath.getPath, value).as[String])
-          .recover( { case jre: JsResultException => throw new ResponseCullingException(s"Invalid path specified for device id: ${devicePath.getPath}", jre) } )
-          .get
+      def extractDeviceIdFieldValue(jsValue: JsValue): Try[String] = {
+        JSONPath.query(devicePath.getPath, jsValue) match {
+          case jsValue: JsNumber =>
+            Success(jsValue.value.toString())
+          case jsValue: JsString =>
+            Success(jsValue.value)
+          case _: JsUndefined =>
+            Failure(new InvalidJsonPathException(s"Invalid path specified for device id: ${devicePath.getPath}"))
+          case _ =>
+            Failure(new InvalidJsonTypeException(s"Invalid JSON type in: ${devicePath.getPath}"))
+        }
+      }
 
-        try {
-          val matcher: Matcher = devicePath.getRegex.getValue.r.pattern.matcher(deviceValue)
+      def parseDeviceId(fieldValue: String): Try[String] = {
+        Try {
+          val matcher = devicePath.getRegex.getValue.r.pattern.matcher(fieldValue)
           if (matcher.matches()) {
-            devicePermissions.exists(_.device == matcher.group(devicePath.getRegex.getCaptureGroup).toInt)
+            matcher.group(devicePath.getRegex.getCaptureGroup)
           } else {
-            throw new ResponseCullingException(s"Regex: ${devicePath.getRegex.getValue} did not match $deviceValue")
+            throw new NonMatchingRegexException(s"Regex: ${devicePath.getRegex.getValue} did not match $fieldValue")
           }
-        } catch {
-          case pse: PatternSyntaxException => throw new ResponseCullingException("Unable to parse regex for device id", pse)
-          case ioobe: IndexOutOfBoundsException => throw new ResponseCullingException("Bad capture group specified", ioobe)
+        } recoverWith {
+          case pse: PatternSyntaxException => Failure(new InvalidRegexException("Unable to parse regex for device id", pse))
+          case ioobe: IndexOutOfBoundsException => Failure(new InvalidCaptureGroupException("Bad capture group specified", ioobe))
```

Comment:
Using `recoverWith` and `flatMap` give you a great way to flatten your exception catching and deal with everything at the end, rather than trying to guess where the exceptions will hit, and having a nested mess. It's super handy once you figure it out.